### ISSUE TITLE
`torch.unique()` messes around with order even if `sorted=False`

### ISSUE THREAD

**Issue by ImahnShekhzadeh** [pytorch/pytorch#105742]
### 🐛 Describe the bug

Hi, 

When I use `torch.unique` without specifying `dim` (in which case it is known that PyTorch will sort the output, cf. [here](https://pytorch.org/docs/stable/generated/torch.unique.html)) and when specifying `sorted=False`, I find that PyTorch still messes around with the order. 

**Code for reproducing:**
```python
import torch 

unique_classes, counts = torch.unique(
    input=torch.tensor([0, 0, 2, 1, 1, 1]).long(), 
    sorted=False, 
    return_counts=True
)
print(unique_classes, counts)
```
yields the output 
```python
(tensor([1, 2, 0]), tensor([3, 1, 2]))
```

**Expected behavior:**:
Since the element `0` appears before `1` or `2` in the input tensor, I would expect `unique_classes` to be 
```python
tensor([0, 2, 1])
```
instead of 
```python
tensor([1, 2, 0])
```

Best, 
Imahn

### Versions

conda 23.3.1
PyTorch 2.0.1 (GPU version)

cc @svekars @carljparker @ezyang @gchanan @zou3519

**Comment by janeyx99** [comment]
It looks like the docs have made a note of this discrepancy and recommends using unique_consecutive to avoid sorting.
![image](https://github.com/pytorch/pytorch/assets/31798555/fed1b0b1-4216-4bc4-961e-eb7ba439f108)\

Would an action item here to be to make the message clearer?

**Comment by ImahnShekhzadeh** [comment]
Hi @janeyx99, the docs state that the sorting happens "when dim is specified". However, in my above code snippet, I did *not* provide `dim`, so it should be `None` by default. Do you see the point?

**Comment by janeyx99** [comment]
Oh I see I had missed the "when dim is specified". This is indeed a discrepancy. Marking tentatively as hi-pri for silent incorrectness.

**Comment by janeyx99** [comment]
After some offline discussion with @albanD, we've concluded this is best approached by fixing the docs and directing people to use unique_consecutive to maintain order.

**Comment by ImahnShekhzadeh** [comment]
Thanks for the summary. If you point out to me how I can fix the docs, I'm happy to do it.

**Comment by albanD** [comment]
That makes sense but I have no idea how you would implement that efficiently though.

**Comment by mingfeima** [comment]
@jiayisunx could you double check the numpy implementation. I remembered that numpy also do a sort on unique so as to make it more efficient.

**Comment by jiayisunx** [comment]
@mingfeima , yes, numpy.unique always returns the sorted unique elements of an array.

**Comment by jingxu10** [comment]
fix in https://github.com/pytorch/pytorch/pull/113420 and https://github.com/pytorch/pytorch/pull/113424

**Comment by AsakusaRinne** [comment]
Hi, it seems that PyTorch has changed the behavior of `Unique` to return sorted values even if `sorted=False` is set? Could anyone give me some explanations about why it is? Though returning sorted values make it easy to have an efficient implementation, it looks a bit weird to just ignore the `sorted=False` parameter from user.

**Comment by svekars** [comment]
looks like this was resolved so closing

**Comment by oiao** [comment]
Sorry to bring this up again, @svekars but I would not consider this resolved, as I consider the [current documentation](https://pytorch.org/docs/stable/generated/torch.unique.html) still confusing:

The note in the docs states that regardless of CPU, CUDA or the `sorted` kwarg, the tensor will be sorted.
I would argue that if `sorted=False` can not be implemented for efficiency reasons, it should be removed altogether. Alternatively, implementing it nonetheless and pointing to the possible performance issues could be another option.

**Comment by aifartist** [comment]
@oiao   Here is some more fun.  From the documentation:

... sort the tensor at the ***BEGINNING*** regardless of the **_sort_** argument...
First of all the arg is named _sorted_ and not _sort_ and then the description of _sorted_ says, not at the **beginning** but:  **before** returning...

I've seen so many issues/confusion with unique() while researching a bug I entered yesterday on wrong results for unique on a 2d tensort(dim=1)
